text/0712-query-params-as-derived-data.md

```diff
@@ -0,0 +1,389 @@
+
+### Map a controller's property to a different query param key
+
+Will need to add a note about getters and other properties in the controller cannot be named the same
+as a query param.
+
+### Default values and (de)serialization
+
+Both default values and deserialization can be handled in getters.
+
+Default value:
+```js
+get category() {
+  return this.router.currentRoute.queryParams.category ?? 'Default Value';
+}
+```
+
+Deserialization:
+```js
+get category() {
+  return parseQP(this.router.currentRoute.queryParams.category);
+}
+```
+`parseQP` defined elsewhere. Receives a string as query params are only ever strings
+
+### Sticky Query Param Values
+
+------------------------------------
+
+#### `{ scope: 'controller' }`
+
+[Demo Ember Twiddle](https://ember-twiddle.com/567b7acf47448cee1f63fcb36e82cd66?openFiles=controllers.articles%5C.js%2C)
+
+This config is not in conflict with derived data.
+
+
+### Default Query Params
+
+Controllers also support default query params, but encourage the use of a
+query param property, which is two-way-bound. To have a one-way dataflow
+from the URL, default query params may look like this:
+
+```js
+export default MyRoute extends Route {
+  @service router;
+
+  async beforeModel({ to: { queryParams }}) {
+    if (!queryParams.category) {
+      this.router.transitionTo({ queryParams: { category: 'default value' }});
+    }
+  }
+}
+```
+
+### Does &lt;Route&gt;#refresh() retain query params?
+
+yes
+
+### Does the route model hook receive query params?
+
+yes, on `transition.to.queryParams`
+
+### How could sticky query params in links handled in user-space?
+
+If sticky query params are managed on a service, a custom `LinkTo` component could
+inject that service and look up if query params have been set for a particular
+route target -- and the implementation could decide if the query params should be
+sticky with or without the dynamic segments. A primitive implementation that uses
+the full href + dynamic segments might look like:
+
+```ts
+const CACHE = new Map<Record<string, string>>();
+
+function getForUrl(url: string) {
+  let existing = CACHE.get(url);
+
+  if (!existing) {
+    CACHE.set(url, {});
+
+    return existing;
+  }
+
+  return existing;
+}
+
+class QueryParamsService extends Service {
+  @service router;
+
+  @action
+  cacheQP(qpName, value) {
+    let cacheForUrl = getForUrl(this.router.currentURL);
+
+    cacheForUrl[qpName] = value;
+  }
+
+  @action
+  getQP(qpName) {
+    return getForUrl(this.router.currentURL)[qpName];
+  }
+
+  @action
+  forUrl(url: string) {
+    return getForUrl(url);
+  }
+}
+```
+
+```ts
+// your custom link component
+interface Args {
+  href: string;
+}
+
+export default class MyLink extends Component<Args> {
+  @service router;
+  @service queryParams;
+
+  get href() {
+    // or this.router.urlFor(...);
+    let { href } = this.args;
+    let qps = stickyQPsToQueryString(this.queryParams.forUrl(href));
+
+    return `${href}?${qps}`;
```

Comment:
This is mixing the URL APIs and string concatenation.

I'd love to see Ember support the Web standard URL APIs for this.

Something like this would gain the benifits of proper URI encoding.

```js
let url = new URL(this.args.href, 'thismessage:/');
url.search = stickyQPsToQueryString(…);
return url.href;
```

Reply:
yeah, the example is out of date -- using `URL` is very good, and I wrote all this before I started using URL, URLSearchParams, etc.

This is all userland code tho, so folks should feel free to use URL :tada: